syntax-sloc/raco-sloc.rkt

```diff
@@ -30,20 +33,117 @@
 (define (missing-sloc src)
   (string-append " N/A" "\t" src))
 
+;; lang-line-match : Pregexp -> Path-String -> Boolean
+(define (lang-line-match? px src)
+  (define lang-line (call-with-input-file src read-#lang))
+  (and lang-line (regexp-match-exact? px lang-line)))
+
+;; read-#lang : Input-Port -> (U #f String)
+(define (read-#lang port)
+  (define end-pos
+    (let* ([lang-in (peeking-input-port port)])
+      (and (procedure? (with-handlers ([exn:fail? (lambda (e) #f)])
+                         (read-language lang-in)))
+           (file-position lang-in))))
+  (define str (and end-pos (read-string end-pos (peeking-input-port port))))
+  (define lang-posn* (and (string? str) (regexp-match-positions* "#lang|#!" str)))
+  (and lang-posn*
+       (string-trim (substring str (cdr (last lang-posn*))))))
+
 (module+ main
   (require racket/cmdline)
+  (define *lang-file-pregexp* (make-parameter #f))
   (command-line
    #:program "syntax-sloc"
-   ;; TODO optional arguments
+   #:once-each
+   [("-l" "--lang")
+    lang-pregexp
+    "Only count files with a matching #lang line"
+    (*lang-file-pregexp* (pregexp lang-pregexp))]
    #:args FILE-OR-DIRECTORY
    (displayln SLOC-HEADER)
+   (define-values (matching-lang? directory-sloc/filter)
+     (let ([px (*lang-file-pregexp*)])
+       (if px
+         (let ([matching-lang? (lambda (src) (lang-line-match? px src))])
+           (values matching-lang?
+                   (lambda (src)
+                     (directory-sloc src #:use-file? matching-lang?))))
+         (values (lambda (x) #t)
+                 directory-sloc))))
```

Comment:
Wouldn't it be better to define `matching-lang?` first and then `directory-sloc/filter` separately, instead of this complicated `define-values` stuff?